client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx

```diff
@@ -0,0 +1,348 @@
+	for ( let i = 0; i < sorted1.length; ++i ) {
+		// If the two lists are sorted and an element differs between the two
+		// at any index, they must not contain exactly the same items
+		// in exactly the same quantities
+		if ( sorted1[ i ] !== sorted2[ i ] ) {
+			return false;
+		}
+	}
+
+	return true;
+};
+
+function useIssueAndAssignLicenses(
+	selectedSite?: { ID: number; domain: string } | null,
+	suggestedProducts: string[] = []
+) {
+	const dispatch = useDispatch();
+	const products = useProductsQuery();
+	const sitesCount = useSelector( getSites ).length;
+
+	const paymentMethodRequired = useSelector( doesPartnerRequireAPaymentMethod );
+
+	const issueLicenses = useIssueLicenses();
+	const assignLicensesToSite = useAssignLicensesToSite( selectedSite?.ID );
+
+	const getLicenseIssuedMessage = useGetLicenseIssuedMessage();
+
+	return useMemo( () => {
+		const isReady = issueLicenses.isReady && assignLicensesToSite.isReady;
+
+		const issueAndAssignLicenses = async ( selectedProducts: string[] ) => {
+			if ( ! isReady || selectedProducts.length === 0 ) {
+				return;
+			}
+
+			dispatch(
+				recordTracksEvent( 'calypso_partner_portal_issue_multiple_licenses_submit', {
+					products: selectedProducts.join( ',' ),
+				} )
+			);
+
+			// We want to know when someone purchases different product(s)
+			// from what we recommend on the dashboard
+			if (
+				suggestedProducts.length > 0 &&
+				! containEquivalentItems( selectedProducts, suggestedProducts )
+			) {
+				dispatch(
+					recordTracksEvent(
+						'calypso_partner_portal_issue_multiple_licenses_changed_selection_after_dashboard_visit'
+					)
+				);
+			}
+
+			const selectedSiteId = selectedSite?.ID;
+			const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+
+			// If we need a payment method, redirect now to have the user enter one
+			if ( paymentMethodRequired ) {
+				const nextStep = addQueryArgs(
+					{
+						products: selectedProducts.join( ',' ),
+						...( selectedSiteId && { site_id: selectedSiteId } ),
+						...( fromDashboard && { source: 'dashboard' } ),
+					},
+					partnerPortalBasePath( '/payment-methods/add' )
+				);
+				return page( nextStep );
+			}
+
+			const issueLicenseResponses = await issueLicenses.issueLicenses( selectedProducts );
+
+			const issuedLicenses = issueLicenseResponses.filter(
+				( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled'
+			);
+			const issuedKeys = issuedLicenses.map( ( { value } ) => value.license_key );
+			const issuedProducts = issuedKeys
+				.map( ( key ) => {
+					const productSlug = key.split( '_' )[ 0 ];
+					return products?.data?.find( ( p ) => p.slug === productSlug );
+				} )
+				.filter( ( p ): p is APIProductFamilyProduct => p !== undefined );
+
+			// Exit early if we don't see any issued licenses matching a product we know
+			if ( issuedProducts.length === 0 ) {
```

Comment:
Honestly? I'm not sure. But since the main goal of this PR was to refactor, I tried to preserve all the existing behavior I could, whether or not I could determine if it made 100% sense.